### PR TITLE
feat: [DHIS2-7310] Add support for warnings in error reports

### PR DESCRIPTION
**Pull request by enricocolasante**

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/report/TrackerWarningReport.java [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+    private final int lineNumber;
+
+    private TrackerErrorCode warningCode;
+
+    private Object mainObject;
+
+    protected int listIndex;
+
+    public TrackerWarningReport(Class<?> mainKlass, String warningMessage, TrackerErrorCode warningCode, int line,
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/report/TrackerWarningReport.java:84]:
```suggestion
    public TrackerWarningReport( Class<?> mainKlass, String warningMessage, TrackerErrorCode warningCode, int line,
```

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/report/TrackerWarningReport.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    {
+        private final List<Object> arguments = new ArrayList<>();
+
+        public TrackerWarningReportBuilder addArg( Object arg )
+        {
+            this.arguments.add( arg );
+            return this;
+        }
+
+        public TrackerWarningReport build(TrackerBundle bundle )
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/report/TrackerWarningReport.java:122]:
```suggestion
        public TrackerWarningReport build( TrackerBundle bundle )
```

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/programrule/ProgramRuleIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+        ProgramRule programRule = createProgramRule( 'A', program );
+        programRuleService.addProgramRule( programRule );
+
+        ProgramRuleAction programRuleActionWarning = createProgramRuleAction( 'A', programRule );
+        programRuleActionWarning.setProgramRuleActionType( ProgramRuleActionType.SHOWWARNING );
+        programRuleActionWarning.setContent( "WARNING" );
+        programRuleActionService.addProgramRuleAction( programRuleActionWarning );
+
+        programRule.getProgramRuleActions().add(programRuleActionWarning);
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/programrule/ProgramRuleIntegrationTest.java:123]:
```suggestion
        programRule.getProgramRuleActions().add( programRuleActionWarning );
```

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/programrule/ProgramRuleIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        assertEquals( TrackerStatus.OK, trackerImportEnrollmentReport.getStatus() );
+        assertFalse( trackerImportEnrollmentReport.getTrackerValidationReport().getWarningReports().isEmpty() );
+    }
+
+    private TrackerImportParams build(TrackerBundleParams params) {
+        return TrackerImportParams.builder()
+                .user( params.getUser() )
+                .importMode( params.getImportMode() )
+                .importStrategy( params.getImportStrategy() )
+            .skipPatternValidation( params.isSkipTextPatternValidation() )
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/programrule/ProgramRuleIntegrationTest.java:158]: Check indentation here.

**Review by stian-sandvold — APPROVED** [review]
Looks good. A couple of codestyle issues :)

**Review by ameenhere — APPROVED** [review]

**Pull request merged**